src/core/common/time.hpp

```diff
@@ -0,0 +1,263 @@
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for time instance.
+ */
+
+#ifndef TIME_HPP_
+#define TIME_HPP_
+
+#include "openthread-core-config.h"
+
+#include <stddef.h>
+#include <stdint.h>
+
+namespace ot {
+
+/**
+ * @addtogroup core-timer
+ *
+ * @brief
+ *   This module includes definitions for the time instance.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This class represents a time instance.
+ *
+ */
+class Time
+{
+public:
+    /**
+     * This constant defines a maximum time duration ensured to be longer than any other duration.
+     *
+     */
+    static const uint32_t kMaxDuration = ~static_cast<uint32_t>(0UL);
+
+    /**
+     * This is the default constructor for a `Time` object.
+     *
+     */
+    Time(void) {}
+
+    /**
+     * This constructor initializes a `Time` object with a given value.
+     *
+     * @param[in] aValue   The numeric time value to initialize the `Time` object.
+     *
+     */
+    explicit Time(uint32_t aValue) { SetValue(aValue); }
+
+    /**
+     * This method gets the numeric time value associated with the `Time` object.
+     *
+     * @returns The numeric `Time` value.
+     *
+     */
+    uint32_t GetValue(void) const { return mValue; }
+
+    /**
+     * This method sets the numeric time value.
+     *
+     * @param[in] aValue   The numeric time value.
+     *
+     */
+    void SetValue(uint32_t aValue) { mValue = aValue; }
+
+    /**
+     * This method calculates the time duration between two `Time` instances.
+     *
+     * @note Expression `(t1 - t2)` returns the duration of the interval starting from `t2` and ending at `t1`. When
+     * calculating the duration, `t2 is assumed to be in the past relative to `t1`. The duration calculation correctly
+     * takes into account the wrapping of numeric value of `Time` instances. The returned value can span the entire
+     * range of the `uint32_t` type.
+     *
+     * @param[in]   aOther  A `Time` instance to subtract from.
+     *
+     * @returns The duration of interval from @p aOther to this `Time` object.
+     *
+     */
+    uint32_t operator-(const Time &aOther) const { return mValue - aOther.mValue; }
```

Comment:
Can this return a signed value? This method disabled using result of comparison, which is not natural. Now we have to compare before `-`, which means `-` will actual be computed twice.

Reply:
That's sort of intentional.

We want to encourage the use of `<` and `>` to determine if one time is before or after the other one.

The goal for `-` operator is to calculate the duration of interval between a _start_ to an _end_ point in time (which then by its definition assumes _start_ is before _end_) also this way we can use the entire `uint32_t` range when calculating interval/duration between two time stamps.